src/devices_models/devices/common/cost_functions.jl

```diff
@@ -186,32 +213,32 @@ function pwl_gencost_sos!(
     spec::AddCostSpec,
     component_name::String,
     cost_data::Vector{NTuple{2, Float64}},
+    time_period::Int,
 )
     var_name = make_variable_name(spec.variable_type, spec.component_type)
-    variable = get_variable(psi_container, var_name)[component_name, :]
+    variable = get_variable(psi_container, var_name)[component_name, time_period]
     settings_ext = get_ext(get_settings(psi_container))
     export_pwl_vars = get_export_pwl_vars(psi_container.settings)
     @debug export_pwl_vars
     total_gen_cost = JuMP.AffExpr(0.0)
 
     if spec.sos_status == NO_VARIABLE
-        time_steps = model_time_steps(psi_container)
-        bin = ones(length(time_steps))
+        bin = 1.0
         @debug("Using Piecewise Linear cost function but no variable/parameter ref for ON status is passed. Default status will be set to online (1.0)")
     elseif spec.sos_status == PARAMETER
         param_key = encode_symbol(OnVariable, string(spec.component_type))
         bin =
-            get_parameter_container(psi_container, param_name).parameter_array[component_name]
+            get_parameter_container(psi_container, param_key).parameter_array[component_name]
         @debug("Using Piecewise Linear cost function with parameter $(param_key)")
     elseif spec.sos_status == VARIABLE
         var_key = make_variable_name(OnVariable, spec.component_type)
-        bin = get_variable(psi_container, var_key)[component_name, :]
+        bin = get_variable(psi_container, var_key)[component_name, time_period]
         @debug("Using Piecewise Linear cost function with variable $(var_key)")
     else
         @assert false
     end
 
-    for (t, var) in enumerate(variable)
+    for var in variable
```

Comment:
You don't need to use a for loop here since the variable will be a single value for the component and time_period